### PR TITLE
Change indexing from [Color][PieceType] to [Piece] to speed up do_move().

### PR DESCRIPTION
**Pull request by syzygy1**
I measure a small speedup in bench, maybe 1%.

The reason for chanching | to + in make_piece (types.h) is that this should allow the compiler to generate the same code for [make_piece(us, pt)] and [us][pt] and perhaps make better use of addressing modes of the CPU. I did not check the generated code to see whether this actually helps, but it should in any event not hurt.

### CODE REVIEW

**Comment by joergoster** [comment]
@syzygy1 When compiling your branch, I get a warning.
syzygy/tbprobe.cpp:25:14: Warnung: Typ von »psq« passt nicht zur ursprünglichen Deklaration
   extern Key psq[COLOR_NB][PIECE_TYPE_NB][SQUARE_NB];
              ^
position.cpp:39:7: Anmerkung: vorher hier deklariert
   Key psq[PIECE_NB][SQUARE_NB];
       ^

**Comment by syzygy1** [comment]
Strange, I don't get that warning even when compiling with -flto. (Without -flto there is no way for the compiler to see that the two definitions in position.cpp are tbprobe.cpp are not identical.)

I'll fix it.

**Comment by erbsenzaehler** [comment]
I fetch your new version and did not get any warnings and it gives a small speedup, too.

archlinux, gcc-6.1, pgo-build:

**Comment by joergoster** [comment]
Now compiles without warnings. Speedup confirmed.
normal builds
```
Result of  30 runs
==================
base =    1516011  +/- 27731
test =    1533551  +/- 25832
diff =     +17540  +/- 15890

speedup        = +0.0116
P(speedup > 0) =  0.9846

CPU: x86_64
```
pgo-builds
```
Result of  30 runs
==================
base =    1790749  +/- 13764
test =    1792375  +/- 13372
diff =      +1626  +/- 5800

speedup        = +0.0009
P(speedup > 0) =  0.7084

CPU: x86_64
```

**Comment by syzygy1** [comment]
Thanks for confirming the measurement.
It is interesting that the speedup is so small on pgo builds. Could it be that this has to do with the fact that some pgo builds are "luckier" than others? (In any case it is impressive how much SF profits from PGO. Off-topic, but is the explanation for this known? Better inlining perhaps?)

There are some more places in Stockfish where lowering the dimension of an array should lead to slightly more efficient code. For example the from-to history table could be addressed by (move & 4095) instead of by [from][to]. (In theory the compiler could figure this out by itself, but I suspect it does not manage to in reality.)

**Comment by lucasart** [comment]
I can measure a small speed-up on my machine. PGO compiles, 64-bit, bmi2: 
```
$ ./pyshbench ../Stockfish/master ../Stockfish/test 20
run       base       test     diff
  1    2772691    2794122   +21431
  2    1782909    1791746    +8837
  3    2695891    2713234   +17343
  4    2769657    2791040   +21383
  5    2771679    2787965   +16286
  6    2774718    2795150   +20432
  7    1733644    1744801   +11157
  8    2767637    2796180   +28543
  9    1732853    1740800    +7947
 10    2775733    2798241   +22508
 11    2776748    2788989   +12241
 12    2755582    2782855   +27273
 13    1732062    1741998    +9936
 14    2765621    2799273   +33652
 15    1736023    1741199    +5176
 16    2776748    2794122   +17374
 17    1733644    1741199    +7555
 18    2769657    2797210   +27553
 19    2756583    2777764   +21181
 20    2775733    2788989   +13256

Result of  20 runs
==================
base =    2457791  +/- 210628
test =    2475344  +/- 213319
diff =     +17553  +/- 3505

speedup        = +0.0071
P(speedup > 0) =  1.0000

CPU: Intel(R) Core(TM) i7-6700 CPU @ 3.40GHz
```

File: src/types.h [hunk 1]
```diff
@@ -353,7 +353,7 @@ inline Square make_square(File f, Rank r) {
 }
 
 inline Piece make_piece(Color c, PieceType pt) {
-  return Piece((c << 3) | pt);
+  return Piece((c << 3) + pt);
```
Comment (lucasart) [inline src/types.h]: can you revert this change, which is totally unrelated ?

Comment (syzygy1) [inline src/types.h]: I already explained why I changed this. It allows the compiler to perform the addition implicitly by an indexed addressing mode instead of being forced to perform an or.

Basically, any code that now implements [us][pt] will also implement [make_piece(us, pt)], because [us][pt] really means [(us << 3) + pt]. So, if the compiler is any good, [make_piece(us, pt)] will never be less efficient than the current code. But if make_piece() uses an OR, the compiler might be forced to generate something less efficient.

I did not look at the assembly output of gcc, but creating more opportunities for optimisation seems like a good thing where it does not decrease readability. And a + is not less readable than an OR, I would say.

But OK, let's have a look at the assembly output. When performing a castling move, do_move() now needs to create a rook of the right color using make_piece(). With an OR in make_piece(), the assembly code (AT&T syntax) contains:
        movl 40(%rsp), %eax  ; %eax = us
        leal 0(,%rax,8), %r11d ; %r11d = 8 * us
        movl %r11d, %eax      ; %eax = 8 * us
        orl $4, %eax             ; %eax = 8 * us | ROOK = make_piece(us, ROOK)
        cmpl %r13d, 48(%rsp) ; compare to and from squares (do_castling())
with a +, the corresponding generated code reads:
        movl 20(%rsp), %eax  ; %eax = us
        cmpl %r13d, 32(%rsp)  ; compare to and from squares (do_castling())
        leal 0(,%rax,8), %r11d  ; %r11d = 8 * us
        leal 4(%r11), %eax      ; %eax = 8 * us + ROOK = make_piece(us, ROOK)
So we already save 1 instruction here. Note that the compiler could also have simply replaced the orl with an addl, if there had not been a more efficient way of translating the addition.

I think the second piece of code could be one instruction shorter by doing leal 4(,%rax,8),%eax, but the intermediate value in %r11 is needed again later. This would not have been possible in the first piece of code, because you can't do an OR by adding a displacement to a leal instruction.

That + in make_piece() leads to more efficient code than OR is confirmed by the fact that do_move allocates 104 bytes on the stack when using + and 136 bytes when using OR. Apparently it reduces register pressure.

So going to back to OR will hurt speed, but some gain from the patch will probably remain. Just let me know.

**Comment by mcostalba** [comment]
Merged with ca6c9f85a5c

I have reworked a bit castling and put/remove helpers to avoid passing redundant arguments and simplified psq init with a nice operator~(Piece pc) overload.

BTW, thanks for the patch.

**Comment by syzygy1** [comment]
The idea of continuing to pass color in put/remove was that we have it in "us" so we don't need to recompute it from "pc". The parameter is of course not really passed as the put/remove functions will be inlined and are effectively macros. Anyway, thanks for merging :)

**Comment by mcostalba** [comment]
@syzygy1 I have seen that sometime we were pre-calculating them anyhow in the caller, like here:

do_castling<false>(us, make_piece(us, KING), make_piece(us, ROOK), from, to, rfrom, rto);

So the optimization is at least dubious, in those cases I prefer the most natural and consistent approach. Moreover, as you said, these inline functions are eventually completely shuffled inside out by the compiler, so no need to bother too much.

**Comment by syzygy1** [comment]
I have tested the committed patch and it is any case not slower than my version. Maybe recalculating the color in put/remove is even faster because it avoids a reload from the stack of "us". So I agree, no need to bother too much.
